### ISSUE TITLE
Python optimization?

### ISSUE THREAD

**Issue by marshallpierce** [wbolster/temporenc-python#22]
I made a [basic benchmark for the python implementation](https://bitbucket.org/marshallpierce/temporenc-python-benchmarks/overview). I just used the first thing I found when searching for "python benchmark"; maybe there are better options?

Anyway, the upshot is that on my machine python generally takes about 100-200 microseconds to serialize or deserialize 100 objects (so a couple of microseconds on average per object, though that's not really how it works). This is probably fine in practice, so I don't want to give the impression that I'm saying this is an urgent problem or anything.

Now for the totally unfair comparison. My rust implementation has been built for speed: it only ever allocates on the stack, is in a language with a fast compiler, etc, so it's not doing the same job as the Python code is. But, it is fast. Encoding and decoding objects (into pre-existing buffers) for the same 100 objects takes between 200 and 1400 **nano**seconds, if my benchmarks are to be believed. 

So, I thought it might be interesting to explore if it's possible to have an alternate back-end for the Python code that uses the Rust code for byte wrangling. I definitely want to keep a painless, easy-to-understand Python implementation for users who can't compile Rust code and for all the interop value of having another independent implementation. On the other hand, if this is something that's interesting to you, and there are things I can do in the Rust codebase (or not do) to make such a thing easy and/or feasible and that don't require regular Rust users to bend over backwards or anything like that, then it seems like it could be fun. Hypothetically the interop between Rust and Python's C API should not be a huge obstacle. I've done Rust/C interop in other projects and it's fine.

There's still more I need to build in the Rust library for this to be feasible, like polymorphic deserialization (where you don't already know what type you're about to deserialize), but I wanted to float the idea now just in case.

**Comment by wbolster** [comment]
wow, thanks for your work. thanks for the write-up as well, you convinced me i should learn some more about rust someday. :)

this actually all sounds very interesting, but tbh the python lib is likely "fast enough" for all use cases. i did write the python code in parallel with designing the format (since "only-text-no-code" specs are pretty much always horrible), and while i paid reasonable attention to performance, i strongly value code simplicity and hence maintainability. (especially since i maintain waaaaay too many open source projects for various things, and all of this has to happen in lost moments...)

btw, perhaps you are emotionally attached to your bitbucket account, but if not, feel free to host temporenc related git repos under the @temporenc umbrella organisation that i just created on github.

**Comment by wbolster** [comment]
> I just used the first thing I found when searching for "python benchmark"; maybe there are better options?

python has the [timeit module](https://docs.python.org/3/library/timeit.html) in the stdlib, and though it's rather simplistic, it requires pretty much no effort to set up.

but my personal almost-no-setup favourite is ipython with its magic `%timeit` helper which does all the dirty work for you.

```shell
pip install ipython
```

then run `ipython` for a fancy repl:

```python
In [1]: import temporenc

In [2]: %timeit temporenc.packb(year=2014, month=10, day=23, hour=12, minute=30, second=30)
The slowest run took 19.06 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 1.4 µs per loop
```

**Comment by marshallpierce** [comment]
Good to know re: timeit and friends. 

I totally understand your priorities for the python impl; I probably wouldn't want to tackle a C Python module without a good reason either.

I prefer BitBucket's tools which is why I put my own things there, but I think it's nice to have sibling libraries all neatly bundled together, so I've moved to https://github.com/temporenc/temporenc-rust.
